[wgpu-hal] Inline RayQuery Support

daniel-keitel opened the pull request:
**Checklist**

- [x] Run `cargo clippy`.
- [x] Run `RUSTFLAGS=--cfg=web_sys_unstable_apis cargo clippy --target wasm32-unknown-unknown` if applicable.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.
- [x] Consider `unsafe_ops_in_unsafe_fn` #3044
- [x] Consider DX12 for API compatibility
- [x] Acceleration Structure Updates

**Future Work**
- [ ] Acceleration Structure Compaction
- [ ] DX 12 Implementation
- [ ] Ray Tracing Pipeline #3607

**Connections**
#1040
#3020 original PR
gfx-rs/naga#2256

**Description**
Basic Vulkan ray query and acceleration structure support.

This pull request is the continuation of @expenses PR #3020.

The Api is designed to work with a DirectX12 Implementation.

![image](https://user-images.githubusercontent.com/6684680/226192254-d8320573-09de-4c4f-9323-da329428cbe2.png)

**Testing**
Tested with example application (ray-traced-triangle)

expenses commented:
Just did a brief look over this again - I think the code is fine but the suggestions from the original PR should probably still be applied.

daniel-keitel commented:
All suggestions from the original PR: https://github.com/gfx-rs/wgpu/pull/3020#pullrequestreview-1139691697
are applied. All other comments where considered and resulted in code changes.

I will look over the API differences between Vulkan and DX12 Before I consider this PR "Ready for review".

daniel-keitel commented:
I made some changes to the Api to improve DX12 Compatibility:
- Single AS build instead of multiple at once
- No support for indirect AS build

And added some features:
- AABB support (untested)
- Support for Geometry Flags
- Support for multiple geometries in single BLAS

In regards to the previously requested changes:
- https://github.com/gfx-rs/wgpu/pull/3020#discussion_r993877518: Surrounding code changed
-  https://github.com/gfx-rs/wgpu/pull/3020#discussion_r993877876: Done
- https://github.com/gfx-rs/wgpu/pull/3020#discussion_r993878614: Done
- https://github.com/gfx-rs/wgpu/pull/3020#discussion_r993879926: We can't really hide it since this structure may be used inside compute shaders. I know of no nice way of handling *C bit fields* in rust. I added some helper methods inside the example, which could be added to core or hal. (At least the struct is identical for Vk and Dx12)
- https://github.com/gfx-rs/wgpu/pull/3020#discussion_r993880556: Done

I have skimmed parts of the Metal specification and it seams fairly compatible,
the main issue I found is that buffer contents are different.
E.g.: Vulkan, DirectX12 and Metal all require  the instance data inside a buffer,
the data structure inside is the same for Vk, Dx but different for Metal.

JMS55 commented:
Hi, super excited to use raytracing! Is there any way I can help with this PR?

RDambrosio016 commented:
> Single AS build instead of multiple at once

Wouldn't it be better to still have this, but just implement it as multiple build calls on dx12?

daniel-keitel commented:
@JMS55, everything is implemented for this PR, only the changelog and some documentation is missing.
I am currently looking into DX12 to implement it next, if you have experience with,  some help would be appreciated, since I have no experience with DX12. 

I am curios, in which additional features beyond this PR are you interested in.

RDambrosio016 commented:
I looked into the DX12 impl a bit and it seems to be really easy with regards to the dx12 code, but a massive pain with regards to the bindings. wgpu-hal has been switching to winapi (and eventually windows-rs), but winapi does not have DXR things because [this pr](https://github.com/retep998/winapi-rs/pull/812) was never merged. I tried reimplementing those bindings in a hacky file inside wgpu-hal, but that is also a pain because it requires a different graphics command list interface (ID3D12GraphicsCommandList4).

JMS55 commented:
@daniel-keitel I'm interested in using ray tracing in [Bevy](https://bevyengine.org) for real time global illumination. So I'm waiting for high level wgpu APIs. In terms of features, compaction seems important for memory savings for static meshes. Beyond that, I'm not familiar enough with RT APIs and performance guidelines to know if there's anything else important missing. I assume we'll have to learn what APIs we need to tweak or expose as we go, once users start experimenting with them.

I don't have much experience with RT or DX12, but I can probably help and learn as I go if there's concrete issues you need help with.

RDambrosio016 commented:
The raytracing pipeline is the other very complex part of raytracing. You basically never want to explicitly sample the BVH from a normal shader. A proper raytracing pipeline with raygen, hit, closesthit, miss, etc shaders is the next big step. That includes a lot of other slightly painful things like SBT records and such,

JMS55 commented:
I did some cursory googling, it looks like ray queries have have similar performance to full ray pipelines? Is this info wrong, or just incomplete?

RDambrosio016 commented:
Yes it shouldn't be a huge difference, but the different shader types have more features, and it allows the GPU to schedule them in a much more custom manner. For example, SER (Shader Execution Reordering) on ada lovelace GPUs. It also hels to separate everything, its much cleaner.

daniel-keitel commented:
@JMS55 Thanks, I will change the API to allow multiple builds at once. I'm interested in using it inside Bevy too.

daniel-keitel commented:
> I did some cursory googling, it looks like ray queries have have similar performance to full ray pipelines? Is this info wrong, or just incomplete?

As far I know it is true for NVIDIA Gpus, AMD ones had performance issues the last time I checked.

But I definitely agree, that we need the rt pipeline, for proper rt support.

The question I am asking myself is, should we implement the low level first (Dx12, compaction, ...),
or should we move to wgpu-core and wgpu?

I couldn't find a clear answer how, we would implement it in the wgpu-core and wgpu crates.
I expect we would simply implement it inside the existing creates and hide it behind a feature flag?

JMS55 commented:
I would implement RT pipelines, and then move onto wgpu-core/wgpu. DX12 can be postponed until Vulkan works, so long as you ensure that the APIs are designed to work well for both VK/DX12, and aren't VK specific. Compaction is really important for memory usage, but not functionality, and can be added later. 

Probably not a feature flag. Maybe put the RT stuff in the [util](https://docs.rs/wgpu/latest/wgpu/util/index.html) module? Up to the maintainers.

daniel-keitel commented:
Everything is implemented for this PR (and somewhat documented).
I will continue by working on the ray-tracing pipeline.

daniel-keitel commented:
@JMS55 are you interested in working together on the next PR?

JMS55 commented:
> @JMS55 are you interested in working together on the next PR?

I would be happy to :). Unfortunately, I don't have my desktop for another week, so I can't test any ray tracing stuff. I can still help though, as it's just wrapping other APIs. I don't need to test it directly to work on it. Just let me know how you want to divide the work.

JMS55 commented on wgpu-hal/src/lib.rs:
```diff
@@ -548,6 +567,18 @@ pub trait CommandEncoder<A: Api>: Send + Sync + fmt::Debug {
+    /// To get the required sizes for the buffer allocations use `get_acceleration_structure_build_sizes` per descriptor
+    /// All buffers must be synchronized externally
+    /// All buffer regions, which are written to may only be passed once per function call,
+    /// with the exertion of updates in the same descriptor.
```
```suggestion
    /// with the exception of updates in the same descriptor.
```

JMS55 commented:
> @JMS55 it would great if we could talk about the Api surface we want to create.

Sure. Does discord messaging work? I can email my discord ID to the email in your user profile.

expenses commented:
@daniel-keitel I have access to a GPU with ray tracing support again, so let me know if I can help in any way to push this forward.

daniel-keitel commented:
@expenses This PR should be "complete" ,general feedback is appreciated, (for #3607 as well).
And It would be great if you would join the discussion about support in wgpu-core, in the matrix thread:
https://matrix.to/#/!FZyQrssSlHEZqrYcOb:matrix.org/$vuQn2Lx30kOuFttuHHTpVZbKcZvCMScAMzTDi8r3s2A?via=matrix.org&via=mozilla.org&via=ralith.com

daniel-keitel commented:
After thinking about the higher level APIs, I will likely change this API to take, iterators instead of slices, to reduce temporary allocations.

daniel-keitel commented:
> After thinking about the higher level APIs, I will likely change this API to take, iterators instead of slices, to reduce temporary allocations.

I will leave it as it is for now, since Iterators over Enums containing (boxed dyn) Iterators, seam kind of wrong to me.
It additionally makes debugging more difficult.

Jaisiero commented:
Hi there!

I downloaded your commit and tried to run the example on Win11 inside this path:
_wgpu\wgpu-hal\examples\ray-traced-triangle_

Gives me this error:
`thread 'main' panicked at 'Selected backend is not supported: InstanceError', wgpu-hal\examples\ray-traced-triangle\main.rs:1080:43`


Am I doing something wrong?

Cheers,
Jaime.

daniel-keitel commented:
@Jaisiero the error handling, isn't very verbose in the example. 
Can you tell me what gpu you have? (I suspect it is incompatible, I will add more verbose error messages, if that's not the case)

Jaisiero commented:
> @Jaisiero the error handling, isn't very verbose in the example. Can you tell me what gpu you have? (I suspect it is incompatible, I will add more verbose error messages, if that's not the case)

After debugging a little, it seems like in main.rs:238 adapters are empty. Maybe I am running it in a wrong way?

daniel-keitel commented:
> > @Jaisiero the error handling, isn't very verbose in the example. Can you tell me what gpu you have? (I suspect it is incompatible, I will add more verbose error messages, if that's not the case)
> 
> After debugging a little, it seems like in main.rs:238 adapters are empty. Maybe I am running it in a wrong way?

@Jaisiero If that's the case and 'halmark' also doesn't work, then the issue is most likely due to wrong features.
Did you start it by executing `cargo run --example ray-traced-triangle`?

Jaisiero commented:
> @Jaisiero the error handling, isn't very verbose in the example. Can you tell me what gpu you have? (I suspect it is incompatible, I will add more verbose error messages, if that's not the case)

I have a RTX 3090. It definitely supports ray tracing but maybe I am not enabling vulkan in my PC. Do I must configure wgpu to use vulkan backend?

Jaisiero commented:
> > > @Jaisiero the error handling, isn't very verbose in the example. Can you tell me what gpu you have? (I suspect it is incompatible, I will add more verbose error messages, if that's not the case)
> > 
> > 
> > After debugging a little, it seems like in main.rs:238 adapters are empty. Maybe I am running it in a wrong way?
> 
> @Jaisiero If that's the case and 'halmark' also doesn't work, then the issue is most likely due to wrong features. Did you start it by executing `cargo run --example ray-traced-triangle`?

Allright, I tried again from the main wgpu folder and now the example works (idk what i did previously).

So from here, the problem I see is that WGSL doesn't support .rgen, .rchit, .rmiss and .rahit. Do you think I could use native GLSL spv shaders to build the ray tracing pipeline? Next step should be to build the Shader Binding Table. This is gonna be a lot of work haha.

Cheers,
Jaime.

daniel-keitel commented:
@Jaisiero I have already implemented the ray-pipeline (with sbt) for wgpu-hal with spv shaders in PR #3607.
My focus currently lies in PR #3631, to add raq-query and acceleration structure support to the wgpu-core and wgpu api,
I will continue working on that next week.

For what do you want to use ray tracing?

Jaisiero commented:
> @Jaisiero I have already implemented the ray-pipeline (with sbt) for wgpu-hal with spv shaders in PR #3607. My focus currently lies in PR #3631, to add raq-query and acceleration structure support to the wgpu-core and wgpu api, I will continue working on that next week.
> 
> For what do you want to use ray tracing?

I would like to build a voxel engine in wgpu and using hardware accel. struct would speed it up! I will need AABB primitives (procedural) for it. Actually I made a primitive one in raw Vulkan. Wouldn't it be great to have hardware accel. RT on web? It should be the first tech to achieve it, right?

expenses commented:
Hey @daniel-keitel, this branch is a little behind so I just merged trunk into it: https://github.com/expenses/wgpu/tree/hal-acceleration-structures, merge commit: https://github.com/expenses/wgpu/commit/fec5e40c125a388f2ebb37cf4cc8827c3d10aa57

JMS55 commented:
@expenses merged trunk into this branch again based off of yours: https://github.com/expenses/wgpu/pull/2. Will try and rebase https://github.com/gfx-rs/wgpu/pull/3631 next.

cwfitzgerald commented:
Could someone tell me how all these MRs are related, I'm having a lot of trouble following them.

JMS55 commented:
This PR is just the wgpu_hal VK backend for RT. No metal, dx12, wpgu-core, or wgpu stuff.

This other PR https://github.com/gfx-rs/wgpu/pull/3631 is building on that and implements RT support at the wgpu-core and wgpu level.

EDIT: BLAS/TLAS support only. No RT pipelines atm. These are meant to be used with inline RT queries, as you picked up on.

cwfitzgerald commented:
Cool - I've renamed them - is this ready for review and landing?

JMS55 commented:
I believe so? I'll leave it up to @daniel-keitel to confirm, though.

daniel-keitel commented:
It is ready for review, after I had a look at the failed tests.

cwfitzgerald commented on wgpu-hal/Cargo.toml:
```diff
@@ -135,6 +135,7 @@ features = ["wgsl-in"]
 cfg-if = "1"
 env_logger = "0.10"
 winit = { version = "0.28.6", features = [ "android-native-activity" ] }   # for "halmark" example
+glam = "0.21.3" # for ray-traced-triangle example
```
Let's sort this alphabetically

cwfitzgerald commented on wgpu-hal/examples/ray-traced-triangle/shader.comp:
```diff
@@ -0,0 +1,44 @@
+#version 460
+#extension GL_EXT_ray_query : enable
```
Why are we using glsl for this? Shouldn't wgsl be supported?

daniel-keitel replied:
It was not yet supported in naga, when I started working on this PR.
I changed It to wgsl.

cwfitzgerald commented on wgpu-hal/examples/ray-traced-triangle/shader.comp:
```diff
@@ -0,0 +1,44 @@
+
+void main() 
+{
+    uvec2 launch_id = gl_GlobalInvocationID.xy;
```
Have some mixed indenting

cwfitzgerald commented on wgpu-hal/src/vulkan/device.rs:
```diff
@@ -2010,6 +2064,223 @@ impl crate::Device<super::Api> for super::Device {
+        &self,
+        acceleration_structure: super::AccelerationStructure,
+    ) {
+        let ray_tracing_functions = match self.shared.extension_fns.ray_tracing {
+            Some(ref functions) => functions,
+            None => panic!("Feature `RAY_TRACING` not enabled"),
+        };
```
This can just be an as_ref().expect() in this and other places

cwfitzgerald commented on wgpu-hal/src/vulkan/command.rs:
```diff
@@ -414,6 +414,269 @@ impl crate::CommandEncoder<super::Api> for super::CommandEncoder {
+            ranges_ptrs.push(&ranges_storage[i]);
+        }
+
+        // let mut geometry_infos =
+        //     Vec::<vk::AccelerationStructureBuildGeometryInfoKHR>::with_capacity(descriptors.len());
+
+        // let mut ranges_vec =
+        //     Vec::<&[vk::AccelerationStructureBuildRangeInfoKHR]>::with_capacity(descriptors.len());
+
+        // let mut ranges_storage =
+        //     Vec::<Vec::<vk::AccelerationStructureBuildRangeInfoKHR>>::with_capacity(descriptors.len());
+
+        // for desc in descriptors {
+        //     let (ranges, geometry_info) = prepare_geometry_info_and_ranges(desc);
+        //     geometry_infos.push(geometry_info);
+        //     ranges_storage.push(ranges);
+
+        // }
+
+        // for i in 0..descriptors.len() {
+        //     ranges_vec.push(&ranges_storage[i]);
+        // }
+
+        // let (ranges, geometry_info) = prepare_geometry_info_and_ranges(descriptors[0]);
```
What's up with these?

daniel-keitel replied:
must have forgotten to remove those

cwfitzgerald commented on wgpu-hal/examples/ray-traced-triangle/main.rs:
```diff
@@ -0,0 +1,1124 @@
+    fn init(window: &winit::window::Window) -> Result<Self, hal::InstanceError> {
+        let instance_desc = hal::InstanceDescriptor {
+            name: "example",
+            flags: if cfg!(debug_assertions) {
+                hal::InstanceFlags::all()
+            } else {
+                hal::InstanceFlags::empty()
+            },
```
This is now wgt::InstanceFlags::default()

cwfitzgerald commented on wgpu-hal/examples/ray-traced-triangle/main.rs:
```diff
@@ -0,0 +1,1124 @@
+            } else {
+                hal::InstanceFlags::empty()
+            },
+            dx12_shader_compiler: wgt::Dx12Compiler::Fxc,
```
We should name Dxc here, as we need it for d3d

cwfitzgerald commented on wgpu-hal/examples/ray-traced-triangle/main.rs:
```diff
@@ -0,0 +1,1124 @@
+        unsafe { cmd_encoder.begin_encoding(Some("init")).unwrap() };
+
+        unsafe {
+            cmd_encoder.place_acceleration_structure_barrier(hal::AccelerationStructureBarrier {
+                usage: hal::AccelerationStructureUses::empty()
+                    ..hal::AccelerationStructureUses::BUILD_OUTPUT,
+            });
```
Why isn't this just a normal barrier?

daniel-keitel replied:
As far as I can tell Vulkan and DX12 require quite different syncronisation, (since dx uses a buffer for the acceleration structure while vulkan has its own resource type).
So this function is in preperation of the DX12 implementation.

cwfitzgerald reviewed (changes requested):
Sorry for taking so long to review this, we'll be better about this going forward.

This looks great! Some smaller nits and needs a rebase, but let's get this in!

cwfitzgerald commented:
(Please re-request my review when changes are addressed!)

JMS55 commented:
Looks like this PR needs rebasing btw. Happy to see more progress on it though, thank you!
